src/coreclr/jit/lsra.cpp

```diff
@@ -5329,34 +5371,139 @@ void LinearScan::allocateRegisters()
             }
             else if ((genRegMask(assignedRegister) & currentRefPosition.registerAssignment) != 0)
             {
-                currentRefPosition.registerAssignment = assignedRegBit;
-                if (!currentInterval->isActive)
+#ifdef TARGET_ARM64
+                if (hasConsecutiveRegister && currentRefPosition.isFirstRefPositionOfConsecutiveRegisters())
                 {
-                    // If we've got an exposed use at the top of a block, the
-                    // interval might not have been active.  Otherwise if it's a use,
-                    // the interval must be active.
-                    if (refType == RefTypeDummyDef)
+                    if (areNextConsecutiveRegistersFree(assignedRegister, currentRefPosition.regCount,
+                                                        currentRefPosition.getInterval()->registerType))
                     {
-                        currentInterval->isActive = true;
-                        assert(getRegisterRecord(assignedRegister)->assignedInterval == currentInterval);
+                        // Current assignedRegister satisfies the consecutive registers requirements
+                        currentRefPosition.registerAssignment = assignedRegBit;
+                        INDEBUG(dumpLsraAllocationEvent(LSRA_EVENT_KEPT_ALLOCATION, currentInterval, assignedRegister));
                     }
                     else
                     {
-                        currentRefPosition.reload = true;
+                        // It doesn't satisfy, so do a copyReg followed by assigning consecutive registers
+                        // to remaining refPosition.
+                        assert((currentRefPosition.refType == RefTypeUse) ||
+                               (currentRefPosition.refType == RefTypeUpperVectorRestore));
+                        regNumber copyReg         = assignCopyReg<true>(&currentRefPosition);
+                        lastAllocatedRefPosition  = &currentRefPosition;
+                        regMaskTP copyRegMask     = getRegMask(copyReg, currentInterval->registerType);
+                        regMaskTP assignedRegMask = getRegMask(assignedRegister, currentInterval->registerType);
+                        bool consecutiveAssigned  = setNextConsecutiveRegisterAssignment(&currentRefPosition, copyReg);
+                        assert(consecutiveAssigned);
+
+                        // For consecutive register, it doesn't matter what the assigned register was.
+                        // We have just assigned it `copyRegMask` and that's the one in-use, and not the
+                        // one that was assigned previously.
+
+                        regsInUseThisLocation |= copyRegMask;
+                        if (currentRefPosition.lastUse)
```

Comment:
I will refactor some of this code to move into a method, because it is used at two places.